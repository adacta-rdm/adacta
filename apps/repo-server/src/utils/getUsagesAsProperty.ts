import { and, eq, gt, gte, isNull, lte, or } from "drizzle-orm";

import type { EntityLoader } from "~/apps/repo-server/src/services/EntityLoader";
import type { DrizzleEntity, DrizzleSchema } from "~/drizzle/DrizzleSchema";
import type { IDeviceId, ISampleId } from "~/lib/database/Ids";

function selectPropertyByValue(Property: DrizzleSchema["Property"], id: IDeviceId | ISampleId) {
	return or(eq(Property.sampleId, id as ISampleId), eq(Property.deviceId, id as IDeviceId)); // Identify the property with the sample/device as value
}

/**
 * Get all usages of a device or sample as a property
 *
 * @param el
 * @param Property
 * @param id
 * @param timeFrame
 * @param includeOverlaps If is set to true, then properties that are only valid for a part of the interval are also taken into account.
 */
export async function getUsagesAsProperty(
	el: EntityLoader,
	Property: DrizzleSchema["Property"],
	id: IDeviceId | ISampleId,
	timeFrame?: { begin?: Date; end?: Date },
	includeOverlaps?: boolean
): Promise<DrizzleEntity<"Property">[]> {
	const conditions = [selectPropertyByValue(Property, id)];
	if (timeFrame) {
		// Only include properties which fully lie within given timeframe
		const end = timeFrame.end;
		const begin = timeFrame.begin;

		if (includeOverlaps) {
			// Checks for properties with  overlap. Should be similar to:
			// (end_prop >= begin || end == null) && begin_prop < end &&
			if (begin) {
				conditions.push(or(isNull(Property.end), gt(Property.end, begin)));
			}

			if (end) {
				conditions.push(and(lte(Property.begin, end)));
			}
		} else {
			if (begin) {
				conditions.push(lte(Property.begin, begin));
			}

			if (end) {
				conditions.push(gte(Property.end, end));
			}
		}
	}

	return el.drizzle
		.select()
		.from(Property)
		.where(and(...conditions));
}
