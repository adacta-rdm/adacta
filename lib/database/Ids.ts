import type { Opaque } from "type-fest";

export type IDeviceDefinitionId = Opaque<string, "DEVICE_DEFINITION">;
export type IDeviceId = Opaque<string, "DEVICE">;
export type IIdPoolId = Opaque<string, "ID_POOL">;
export type IImportPresetId = Opaque<string, "IMPORT_PRESET">;
export type INameCompositionId = Opaque<string, "NAME_COMPOSITION">;
export type INameCompositionVariableId = Opaque<string, "NAME_COMPOSITION_VARIABLE">;
export type INoteId = Opaque<string, "NOTE">;
export type IProjectId = Opaque<string, "PROJECT">;
export type IPropertyId = Opaque<string, "PROPERTY">;
export type IRepositorySettingId = Opaque<string, "REPOSITORY_SETTING">;
export type IResourceId = Opaque<string, "RESOURCE">;
export type ISampleId = Opaque<string, "SAMPLE">;
export type ISampleRelationId = Opaque<string, "SAMPLE_RELATION">;
export type ITransformationId = Opaque<string, "TRANSFORMATION">;
export type IUserId = Opaque<string, "USER">;

/**
 * Maps the GraphQL __typename to their respective ID types.
 * Note: This is only required for types that are not part of CouchDB or types that are named
 * differently than the "type" field of the CouchDB document.
 */
export interface IIdTypeMap {
	// Register types from auth-server (prisma)
	User: IUserId;
	RemoteRepo: IRepositorySettingId;

	// Register alias types from auth-server
	AuthenticatedUser: IUserId;

	// Register types from repo-server (MikroORM)
	Note: INoteId;

	// Register alias types from repo-server
	ResourceGeneric: IResourceId;
	ResourceImage: IResourceId;
	ResourceImportPreset: IResourceId;
	ResourceTabularData: IResourceId;

	// Register MikroORM types
	NameComposition: INameCompositionId;
	NameCompositionVariable: INameCompositionVariableId;
	NameCompositionVariableStatic: INameCompositionVariableId;
	NameCompositionVariableVariable: INameCompositionVariableId;
}
