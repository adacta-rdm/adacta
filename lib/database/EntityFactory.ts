import type { IUserId } from "./Ids";

import { AuthenticatedUserInfo } from "~/apps/repo-server/src/graphql/AuthenticatedUserInfo";
import { entityId } from "~/apps/repo-server/src/utils/entityId";
import type { DrizzleEntity, DrizzleEntityNameId, Metadata } from "~/drizzle/DrizzleSchema";
import { entityByTypeId } from "~/drizzle/DrizzleSchema";
import { Service } from "~/lib/serviceContainer/ServiceContainer";

@Service(AuthenticatedUserInfo)
export class EntityFactory {
	static id<T extends DrizzleEntityNameId>(e: T): DrizzleEntity<T>["id"] {
		const entry = Object.entries(entityByTypeId).find(([, v]) => v === e);
		if (!entry) throw new Error(`No entity with typeid ${e}`);
		return entityId(entry[0]) as DrizzleEntity<T>["id"];
	}

	static create<T extends "User", TProps extends EntityProps<"User">>(
		entityName: T,
		props: TProps
	): TProps & { id: DrizzleEntity<T>["id"] };
	static create<T extends DrizzleEntityNameId, TProps extends EntityProps<T>>(
		entityName: T,
		props: TProps,
		creatorId: IUserId
	): TProps & Metadata & { id: DrizzleEntity<T>["id"] };
	static create<T extends DrizzleEntityNameId, TProps extends EntityProps<T>>(
		entityName: T,
		props: TProps,
		creatorId?: IUserId
	) {
		const o = {
			...props,
			id: EntityFactory.id(entityName),
		};

		if (entityName === "User") return o;

		if (!creatorId) {
			throw new Error(`Missing required argument 'creatorId' for entity ${entityName}`);
		}

		return Object.assign(o, EntityFactory.metadata(creatorId));
	}

	constructor({ userId }: AuthenticatedUserInfo) {
		this.userId = userId;
	}

	create<T extends DrizzleEntityNameId, TProps extends EntityProps<T>>(
		entityName: T,
		props: TProps
	) {
		return EntityFactory.create(entityName, props, this.userId);
	}

	private userId: IUserId;

	private static metadata(metadataCreatorId: IUserId) {
		return {
			metadataCreatorId,
			metadataCreationTimestamp: new Date(),
			metadataDeletedAt: null,
		} satisfies Metadata;
	}
}

type EntityProps<T extends DrizzleEntityNameId> = Omit<
	DrizzleEntity<T>,
	// Remove the id/metadata fields as these are populated by the factory
	"metadataCreatorId" | "metadataCreationTimestamp" | "metadataDeletedAt" | "id"
>;
