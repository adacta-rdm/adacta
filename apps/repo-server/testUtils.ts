import { PGlite } from "@electric-sql/pglite";
import { assertDefined } from "@omegadot/assert";
import { PgDialect } from "drizzle-orm/pg-core";
import { drizzle as pgliteDrizzle } from "drizzle-orm/pglite/driver";
import type { SQL } from "drizzle-orm/sql/sql";

import { AuthenticatedUserInfo } from "~/apps/repo-server/src/graphql/AuthenticatedUserInfo";
import { EntityLoader } from "~/apps/repo-server/src/services/EntityLoader";
import { RepositoryManagerPostgres } from "~/apps/repo-server/src/services/RepositoryManagerPostgres";
import { isEntityId } from "~/apps/repo-server/src/utils/isEntityId";
import type { DrizzleEntity } from "~/drizzle/DrizzleSchema";
import { entityByTypeId } from "~/drizzle/DrizzleSchema";
import { EntityFactory } from "~/lib/database/EntityFactory";
import type { IDeviceId, ISampleId, IUserId } from "~/lib/database/Ids";
import { SilentLogger } from "~/lib/logger/SilentLogger";
import { ServiceContainer } from "~/lib/serviceContainer/ServiceContainer";
import { uuid } from "~/lib/uuid";

/**
 * Wraps a simple integer into a date
 * @param day
 */
export function integerToDate(day: number) {
	return new Date(2020, 0, day);
}

/**
 * Unwraps the date back into an integer (reverse action of `integerToDate`)
 * @param a
 */
export function dateToInteger(a: Date | undefined | null) {
	if (!a) {
		return "now";
	}
	return (a.getTime() - integerToDate(1).getTime()) / 60 / 60 / 24 / 1000 + 1;
}

export class TestDeviceCreator {
	device?: DrizzleEntity<"Device">;
	userId: IUserId;
	properties: DrizzleEntity<"Property">[] = [];

	constructor(user: IUserId) {
		this.userId = user;
	}

	public addDevice(definition: DrizzleEntity<"DeviceDefinition">, deviceName: string) {
		this.device = EntityFactory.create(
			"Device",
			{
				couchId: null,
				shortId: null,
				name: deviceName,
				definitionId: definition.id,
				specifications: [],
				imageResourceIds: [],
				setupDescription: [],
			},
			this.userId
		);
		// Reset collected properties which were for a different device
		this.properties = [];
	}

	public addProperty(
		name: string,
		value: IDeviceId | ISampleId,
		begin: Date,
		end: Date | null = null
	) {
		assertDefined(this.device, "addDevice() needs to be called before calling addProperty()");

		this.properties.push(
			EntityFactory.create(
				"Property",
				{
					name,
					begin,
					end,
					ownerDeviceId: this.device.id,
					deviceId: isEntityId(value, "Device") ? value : null,
					sampleId: isEntityId(value, "Sample") ? value : null,
				},
				this.userId
			)
		);
	}

	public getEntities(): [DrizzleEntity<"Device">, ...DrizzleEntity<"Property">[]] {
		assertDefined(this.device, "TestDeviceCreator: Device not defined");
		return [this.device, ...this.properties];
	}
}

export const createTestDb = async () => {
	const sc = new ServiceContainer();

	const drizzle = pgliteDrizzle(new PGlite());
	const rmp = sc.set(new RepositoryManagerPostgres({} as any));
	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
	rmp.db = () => drizzle as any;
	const el = sc.set(new EntityLoader(drizzle as any));
	await rmp.migrate();
	await rmp.createRepository("test");
	const schema = sc.set(rmp.schema("test"));

	const counter = new Map<string, number>();

	// Override EntityFactory.id to generate deterministic ids
	EntityFactory.id = (e) => {
		const entry = Object.entries(entityByTypeId).find(([, v]) => v === e);
		if (!entry) throw new Error(`No entity with typeid ${e}`);
		const hex = Number(entry[0]).toString(16).padStart(2, "0");

		const count = counter.get(hex) ?? 0;
		counter.set(hex, count + 1);

		const id = `${count}`.padStart(12, "0");
		return `00000000-0000-00${hex}-0000-${id}` as DrizzleEntity<typeof e>["id"];
		// 00000000-0000-x000-y000-000000000000
		//               14   19
	};

	const user = EntityFactory.create("User", {
		mikroORMId: uuid(),
		firstName: "Test",
		lastName: "User",
		email: "test.user@example.com",
		passwordHash: "password",
		salt: "salt",
		locale: "en",
		dateStyle: "short",
		timeStyle: "short",
	});
	await drizzle.insert(schema.User).values(user);

	sc.set(new AuthenticatedUserInfo(user.id, "token"));
	sc.set(new SilentLogger());

	return { drizzle, el, schema, user, sc };
};

// Test UUID use a custom format, where the last part is a simple counter
export function testUUIDtoInteger(id: string) {
	const n = id.lastIndexOf("-");
	return `${Number(id.substring(n + 1))}`;
}

export function debugRawQuery(query: SQL<unknown>) {
	const pgDialect = new PgDialect();
	// eslint-disable-next-line no-console
	console.log(pgDialect.sqlToQuery(query.inlineParams()).sql);
}
