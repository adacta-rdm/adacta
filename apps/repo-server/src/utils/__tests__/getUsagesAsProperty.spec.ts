import { describe, test, expect } from "vitest";

import { createTestDb, dateToInteger, integerToDate } from "../../../testUtils";
import { getUsagesAsProperty } from "../getUsagesAsProperty";

import type { DrizzleEntity } from "~/drizzle/DrizzleSchema";
import { EntityFactory } from "~/lib/database/EntityFactory";
import type { IDeviceDefinitionId, IDeviceId, IUserId } from "~/lib/database/Ids";

const deviceNameById = new Map<IDeviceId, string>();

const createDevice = (
	uniqueName: string,
	deviceDefinitionId: IDeviceDefinitionId,
	userId: IUserId
) => {
	const doc = EntityFactory.create(
		"Device",
		{
			name: uniqueName,
			definitionId: deviceDefinitionId,
			imageResourceIds: [],
			specifications: [],
			setupDescription: [],
		},
		userId
	);

	deviceNameById.set(doc.id, uniqueName);

	return doc;
};

const sanitize = (docs: DrizzleEntity<"Property">[]) => {
	return docs
		.sort((a, b) => {
			return a.begin.getTime() - b.begin.getTime();
		})
		.map((d) => {
			return {
				...d,
				id: undefined,
				ownerDeviceId: deviceNameById.get(d.ownerDeviceId),
				deviceId: deviceNameById.get(d.deviceId as IDeviceId),
				begin: dateToInteger(d.begin),
				end: dateToInteger(d.end),
			};
		});
};

describe("getUsagesAsProperty", () => {
	describe("initial test", async () => {
		const { el, schema, user } = await createTestDb();
		const { Property } = schema;

		const userId = user.id;

		const deviceDefinition = EntityFactory.create(
			"DeviceDefinition",
			{
				name: "Test",
				parentDeviceDefinitionIds: [],
				imageResourceIds: [],
				acceptsUnit: [],
				specifications: [],
				couchId: null,
			},
			user.id
		);
		await el.insert(schema.DeviceDefinition, deviceDefinition);

		const root = createDevice("Root-Device", deviceDefinition.id, userId);
		const c1 = createDevice("Child1", deviceDefinition.id, userId);
		const c2 = createDevice("Child2", deviceDefinition.id, userId);
		const c3 = createDevice("Child3", deviceDefinition.id, userId);

		await el.insert(schema.Device, root);
		await el.insert(schema.Device, c1);
		await el.insert(schema.Device, c2);
		await el.insert(schema.Device, c3);

		await el.insert(
			schema.Property,
			EntityFactory.create(
				"Property",
				{
					ownerDeviceId: root.id,
					name: "device1",
					begin: integerToDate(0),
					end: integerToDate(3),
					deviceId: c1.id,
				},
				userId
			)
		);

		await el.insert(
			schema.Property,
			EntityFactory.create(
				"Property",
				{
					ownerDeviceId: root.id,
					name: "device1",
					begin: integerToDate(2),
					end: integerToDate(5),
					deviceId: c1.id,
				},
				userId
			)
		);

		await el.insert(
			schema.Property,
			EntityFactory.create(
				"Property",
				{
					ownerDeviceId: root.id,
					name: "device1",
					begin: integerToDate(5),
					end: integerToDate(10),
					deviceId: c1.id,
				},
				userId
			)
		);

		await el.insert(
			schema.Property,
			EntityFactory.create(
				"Property",
				{
					ownerDeviceId: root.id,
					name: "device2",
					begin: integerToDate(5),
					end: undefined,
					deviceId: c2.id,
				},
				userId
			)
		);

		test.each([
			{ id: c1.id, begin: 0, end: 3 },
			{ id: c1.id, begin: 2, end: 5 },
			{ id: c1.id, begin: 2, end: 10 },
			{ id: c2.id, begin: 2, end: 10 },
			{ id: c2.id, begin: 2, end: null },
			{ id: c1.id, begin: 5, end: 10 },
		])("includeOverlaps: %o", async ({ id, begin, end }) => {
			const args = (includeOverlaps: boolean) =>
				[
					el,
					Property,
					id,
					{
						begin: integerToDate(begin),
						end: typeof end === "number" ? integerToDate(end) : undefined,
					},
					includeOverlaps,
				] as const;

			expect(sanitize(await getUsagesAsProperty(...args(true)))).toMatchSnapshot();
			expect(sanitize(await getUsagesAsProperty(...args(false)))).toMatchSnapshot();
		});
	});
});
