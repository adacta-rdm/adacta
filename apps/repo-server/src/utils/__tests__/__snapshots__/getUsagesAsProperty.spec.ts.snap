// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000001', begin: +0, end: 3 } 1`] = `
[
  {
    "begin": 0,
    "couchId": null,
    "deviceId": "Child1",
    "end": 3,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
  {
    "begin": 2,
    "couchId": null,
    "deviceId": "Child1",
    "end": 5,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
]
`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000001', begin: +0, end: 3 } 2`] = `
[
  {
    "begin": 0,
    "couchId": null,
    "deviceId": "Child1",
    "end": 3,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
]
`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000001', begin: 2, end: 5 } 1`] = `
[
  {
    "begin": 0,
    "couchId": null,
    "deviceId": "Child1",
    "end": 3,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
  {
    "begin": 2,
    "couchId": null,
    "deviceId": "Child1",
    "end": 5,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
  {
    "begin": 5,
    "couchId": null,
    "deviceId": "Child1",
    "end": 10,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
]
`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000001', begin: 2, end: 5 } 2`] = `
[
  {
    "begin": 2,
    "couchId": null,
    "deviceId": "Child1",
    "end": 5,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
]
`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000001', begin: 2, end: 10 } 1`] = `
[
  {
    "begin": 0,
    "couchId": null,
    "deviceId": "Child1",
    "end": 3,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
  {
    "begin": 2,
    "couchId": null,
    "deviceId": "Child1",
    "end": 5,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
  {
    "begin": 5,
    "couchId": null,
    "deviceId": "Child1",
    "end": 10,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
]
`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000001', begin: 2, end: 10 } 2`] = `[]`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000001', begin: 5, end: 10 } 1`] = `
[
  {
    "begin": 5,
    "couchId": null,
    "deviceId": "Child1",
    "end": 10,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
]
`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000001', begin: 5, end: 10 } 2`] = `
[
  {
    "begin": 5,
    "couchId": null,
    "deviceId": "Child1",
    "end": 10,
    "id": undefined,
    "name": "device1",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
]
`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000002', begin: 2, end: 10 } 1`] = `
[
  {
    "begin": 5,
    "couchId": null,
    "deviceId": "Child2",
    "end": "now",
    "id": undefined,
    "name": "device2",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
]
`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000002', begin: 2, end: 10 } 2`] = `[]`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000002', begin: 2, end: null } 1`] = `
[
  {
    "begin": 5,
    "couchId": null,
    "deviceId": "Child2",
    "end": "now",
    "id": undefined,
    "name": "device2",
    "ownerDeviceId": "Root-Device",
    "sampleId": null,
  },
]
`;

exports[`getUsagesAsProperty > initial test > includeOverlaps: { id: '00000000-0000-0007-0000-000000000002', begin: 2, end: null } 2`] = `[]`;
